File: output_20260318_091315/qwen3/c/c/Struct/spouse.c
/*@ predicate person(struct person* p; struct person* s) =
  p != 0 &*&
  malloc_block_person(p) &*&
  p->spouse |-> s;
@*/

/*@ predicate married(struct person* p, struct person* q) =
  person(p, q) &*& person(q, p);
@*/

struct person {
  char* name;
  struct person* spouse;
};

struct person *create_person()
//@ requires true;
//@ ensures person(result, 0);
{
  struct person *p = malloc(sizeof(struct person));
  if(p == 0) abort();
  p->spouse = 0;
  
  return p;
}

void marry(struct person *this, struct person *other)
//@ requires person(this, 0) &*& person(other, 0);
//@ ensures married(this, other);
{
  this->spouse = other;
  other->spouse = this;
}

struct person* person_get_spouse(struct person* this)
//@ requires person(this, ?s);
//@ ensures person(this, s) &*& result == s;
{
  return this->spouse;
}

void divorce(struct person* this)
//@ requires married(this, ?other);
//@ ensures person(this, 0) &*& person(other, 0);
{
  this->spouse->spouse = 0;
  this->spouse = 0;
}

void die(struct person *this)
//@ requires person(this, ?s) &*& (s == 0 || person(s, this));
//@ ensures s == 0 ? true : person(s, 0);
{
  if(this->spouse != 0) {
    this->spouse->spouse = 0;
  }
  free(this); 
}

int main() 
//@ requires true;
//@ ensures true;
{
  struct person* alice = create_person();
  struct person* bob = create_person();
  struct person* eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  die(alice);
  die(bob);
  die(eve);
  return 0;
}

File: output_20260318_123118/qwen3/java/Linklist/Spouse.java
/*@ predicate spouse_of(Person p, Person s) = p.spouse |-> s; @*/

class Person {

    protected Person spouse;

    //@ requires true;
    //@ ensures true;
    public void spouse_symm()
    {
    }

    //@ requires true;
    //@ ensures spouse_of(this, null);
    public Person()
    {
    }
    
    //@ requires spouse_of(this, ?s);
    //@ ensures result == s &*& spouse_of(this, s);
    public Person getSpouse()
    {
        return spouse;
    }
    
    //@ requires spouse_of(this, null) &*& spouse_of(other, null);
    //@ ensures spouse_of(this, other) &*& spouse_of(other, this);
    protected void setSpouse(Person other)
    {
        spouse = other;
        other.spouse = this;
    }
    
    //@ requires spouse_of(this, ?s) &*& s != null &*& spouse_of(s, this);
    //@ ensures spouse_of(this, null) &*& spouse_of(s, null);
    protected void clearSpouse()
    {
        spouse.spouse = null;
        spouse = null;
    }
    
    //@ requires spouse_of(this, null) &*& spouse_of(other, null);
    //@ ensures spouse_of(this, other) &*& spouse_of(other, this);
    void marry(Person other)
    {
        other.setSpouse(this);
    }
    
    //@ requires spouse_of(this, ?s) &*& s != null &*& spouse_of(s, this);
    //@ ensures spouse_of(this, null) &*& spouse_of(s, null);
    void divorce()
    {
        spouse.clearSpouse();
    }

}

class Program {

    //@ requires true;
    //@ ensures true;
    public static void main(String[] args)
    {
        Person a = new Person();
        Person b = new Person();
        a.marry(b);
        b.divorce();
    }

}